Feature/zero config setup

IvanGrginovInf opened the pull request:
Still a **WIP**, opening a PR so this is all tracked.

Updated setup process for boilerplate that should be as simple and intuitive as possible while working on all platforms.

Currently it's 1 script (`npm run setup`) for setting everything up, after which user can follow the manual WordPress wizard install (by going to website's local url) or try the automatic setup depending on his local dev setup.

Would appreciate any feedback on intuitiveness of this.

Currently done
---
* Reworked the setup.sh script so that it's now done in JS (no more bash)
* `rename.js`now works no matter how many times you run it (saves old state to `theme-manifest.json`)
* Added a default `theme-manifest.json` with initial values we replace using `rename.js`

To do
---
* Output user's pass when creating him
* Add some validation to user input (for example: only allow lowercase and hyphens for package name, etc)
* Make the installation pretty ( [chalk](https://www.npmjs.com/package/chalk), some character art, emojis and removing unnecessary output )
* Test on Windows (doesn't use bash anymore so it should be all good, also running `composer` using `npx`)
* Create a Vagrant setup script

dingo-d commented on bin/setup-wp.js:
```diff
@@ -0,0 +1,168 @@
+  dbPass: "wp",
+  wpUser: "ivan.grginov",
+  wpPass: "wp-pass",
+  siteName: "ImeStranice"
```
This is a test or? If any default text will be placed it should be in English :)

IvanGrginovInf replied:
Temp stuff, there's really no need to set any defaults here, will remove. 👍

dingo-d commented on wp-info.json:
```diff
@@ -0,0 +1,7 @@
+  "db-name": "infinum_boilerplate_dev",
+  "db-user": "wp",
+  "db-pass": "wp",
+  "wp-user": "ivan.grginov",
```
I'm fine with having your name, but it would be safer to use something totally generic, like `infinum.user`

IvanGrginovInf commented:
Changes in last push
---
* Added the WP install part after the initial setup (`wp core install`, `wp config create` and `wp theme activate`) depending on the user's local dev environment
* It's now possible to have a working install (with DB and everything) with just 1 call `npm run setup` on VVV (still need to implement the wp installation part for other local setup - as it is now it just tells you to do it manually by going to your local website url and you will see the WP's 5-min wizard setup)
* Reworked the update / renaming functions / scripts so they're all `async`(mostly just so they work with [ora](https://www.npmjs.com/package/ora))
* Reworked the rename script **completely** with new styling

To Do
---
* Add "Create database if it doesn't exist" part before `wp config create`
* Add some basic compatibility check for setups other than VVV (for example check if we're able to connect to `mysql` using "best guess" credentials for a particular setup - already using this with VVV - If not, prompt user and check again)
* Skip setup steps if they're not needed (for example `wp core download`, as it is now if wp core exists the script will exit with an error - ideally it would just skip to the next part)
* Check if search / replace is working with hidden files

dingo-d commented on README.md:
```diff
@@ -12,6 +12,8 @@ This repository contains all the tools you need to start building a modern WordP
 
 First you need to install WordPress locally, using any of the local development environment you prefer. You can use XAMPP, MAMP, WAMP, VVV, Docker or Laravel Valet. You'll also need to have [Node.js](https://nodejs.org/en/), [Composer](https://getcomposer.org/) and [WP-CLI](https://wp-cli.org/) installed.
 
+**Note regarding `Windows` - If you're installing this on `Windows` and you're using `XAMPP`, use `XAMPP` shell (Control Panel -> Shell) rather than `cmd` **
```
This isn't rendered as bold. A better approach imo would be to add this as a subheading like: 

#### Windows installation

If you're installing this on Windows, and you are using `XAMPP`, use `XAMPP` shell (Control Panel -> Shell) rather than the Windows command line prompt.

IvanGrginovInf replied:
Changed this a bit so it no longer needs to be bolded.

dingo-d commented on README.md:
```diff
@@ -66,18 +68,18 @@ Once you're done with the above setup (no matter which method you used), run `np
 
```
At the top where the `## Initial Setup` is, we should add a note that if you are running from the vagrant ssh, these are all installed (node, composer and wp-cli)

dingo-d replied:
Also for the manual part, besides running `npm start`, shouldn't `composer install`be ran as well? Line 41

IvanGrginovInf replied:
You still need `wp-cli` for vagrant (since `wp core download` is run outside vagrant). And you don't actually need composer regardless (at least not for installation) since we're running it using `npx`.

As for composer install, that is done just after renaming (actually it's `update` not `install`). (1. `npm install`, 2. renaming, 3. `npx composer update`, 4. `wp core donwload`)

dingo-d commented on README.md:
```diff
@@ -66,18 +68,18 @@ Once you're done with the above setup (no matter which method you used), run `np
     // Must be set.
     // Possible options are develop, staging and production.
-    define( 'INF_ENV', 'develop' );
+    define( 'SF_ENV', 'develop' );
```
`SF_ENV` is placed to denote already replaced env variable?

dingo-d commented on README.md:
```diff
@@ -121,13 +123,13 @@ Once you're done with the above setup (no matter which method you used), run `np
     Checking theme for possible violations:
```
I'll create the composer scripts that one can run to check the phpcs, once I do this we'll update this part. Aliasing is one extra step, and composer is already there :)

IvanGrginovInf replied:
Great! 👍

dingo-d commented on README.md:
````diff
@@ -121,13 +123,13 @@ Once you're done with the above setup (no matter which method you used), run `np
 
     ```bash
-    wpcs wp-content/themes/inf_theme
+    wpcs wp-content/themes/f
````
Why is this replaced? Also the above will apply with composer scripts.

dingo-d commented on README.md:
````diff
@@ -121,13 +123,13 @@ Once you're done with the above setup (no matter which method you used), run `np
 
     ```bash
-    wpcbf wp-content/themes/inf_theme
+    wpcbf wp-content/themes/f
````
⬆️

dingo-d commented on README.md:
````diff
@@ -121,13 +123,13 @@ Once you're done with the above setup (no matter which method you used), run `np
+    wpcbf wp-content/themes/f
     ```
 
 ## Goint to Production
````
The two scripts should be `build.sh` and `deploy.sh` right?

dingo-d replied:
I couldn't mark the lines below, the readme has 2x `bash bin/build.sh` mentioned for build and deploy. And it should be `bash bin/build.sh` and `bash bin/deploy.sh` 🙂

IvanGrginovInf replied:
They're currently the same so I'm guessing that's why it was written like so. Will change tho. 👍

dingo-d commented on bin/files.js:
```diff
@@ -35,35 +35,90 @@ exports.readManifest = (key) => {
 
-  try {
-    const changes = replace.sync(options);
+  return new Promise((resolve) => {
```
Above you used `async` so how come you used a `Promise` here? Couldn't this be also made using `async/await`?

IvanGrginovInf replied:
It was a c/p (first thing I converted to `async`) and when it worked, I didn't bother with it anymore later. Will rework to async. 👍

dingo-d commented on bin/output.js:
```diff
@@ -0,0 +1,98 @@
+  console.log(chalk.red('    ___  ____ _ _    ____ ____ ___       ____ ___ ____ '));
+  console.log(chalk.red('    |__| |  | | |    |___ |__/ |__| |    |__|  |  |___ '));
+  console.log(chalk.red('    |__| |__| | |___ |___ |  \\ |    |___ |  |  |  |___ '));
+  console.log(chalk.red(''));
```
ASCII art :D Cool

dingo-d commented on package.json:
```diff
@@ -51,13 +51,17 @@
   "dependencies": {
     "@babel/polyfill": "^7.0.0",
     "autoprefixer": "^9.1.3",
+    "chalk": "^2.4.1",
```
Shouldn't this be in the `devDependencies`, it's not meant to be used in production, or bundled in the final build, right?

dingo-d commented on bin/wp-installer.js:
```diff
@@ -0,0 +1,267 @@
+ * @return string
+ */
+const promptVmDir = () => output.prompt({
+  label: '1. Please enter your project\'s \'vm_dir\' VVV path (by default it\'s /srv/www/PROJECTNAME/):',
```
This line is misleading, remove the last `/` from the `PROJECTNAME` 😉 

I added the last line and I got error

```bash
Error: Command failed: vagrant ssh -- -t 'cd src/www/clients/test//public_html/;'
bash: line 0: cd: src/www/clients/test//public_html/: No such file or directory
```

IvanGrginovInf replied:
You're completely right, I just got so used to it I ignored the instructions. :) anyway, will change the instructions + add a trim function if the user adds the extra `/` anyway

IvanGrginovInf commented:
Fixed all the issues. Also @iruzevic can you just look at the `build.sh`/ `deploy.sh`in README ([this commit](https://github.com/infinum/wp-boilerplate/pull/115/commits/e06eac9014b4696397ab88cb774330c64385b428)) if I've changed it correctly.

Other than that I've separated the `rename`script (`npm run rename`) so it can be run whenever. Works as long as info in `theme-manifest.json` (needles) is correct. Added a warning regarding that before prompting.

I think it could be merged , what do you think? @iruzevic @dingo-d 

There are still a few things on my to-do regarding the setup (such as automatic wp install setup for more environments other than `VVV` but these can be implemented later.

dingo-d commented:
Few minor things that are bothering my OCD (please don't hate me 😆)

When you start the setup, the icons are glued to the text, also the cursor is 1 space away when typing which seems kinda odd like I wrote something wrong 😬 

<img width="617" alt="screenshot 2018-10-09 at 18 03 53" src="https://user-images.githubusercontent.com/8638515/46682428-cb1a8280-cbed-11e8-87ea-c48fd5be9cb9.png">

Also I tried to create a vagrant install and failed

<img width="757" alt="screenshot 2018-10-09 at 18 10 26" src="https://user-images.githubusercontent.com/8638515/46682727-a377ea00-cbee-11e8-8dab-5df9c4fb5c11.png">

Now there seems to be some discrepancies between the text in the instructions and what happened.
You wrote `Please enter database name (make sure it exists and is empty):`

I had a db in the vagrant and i emptied it, but I got an error that said

> Can't create database 'wordpress_internal'; database exists

So, something is a miss here 🙂 

Also, from a11y point of view this seems a bit unreadable, and could provide a problem to the people who have impaired vision. 

<img width="616" alt="screenshot 2018-10-09 at 18 13 07" src="https://user-images.githubusercontent.com/8638515/46682893-0ff2e900-cbef-11e8-9acb-ecc9a35defac.png">

Granted, this could be only for my setup, which is totally not a11y friendly. I presume that people who have such problems would have high contrast terminal settings.

This is not a blocker for a release imo, but it would be great if we would make some kind of call to action for users to test, especially if they are visually impaired 🙂 

All in all I love this and if these small issues are fixed, I'm for merging this to develop, testing and releasing the v3 :tada:

Awesome work!

IvanGrginovInf commented:
Changes since your last review:
---
* Separated the wp-setup part (`npm run setup-wp`)
* Added a **_Custom local development server_** setup. Not really tested since I don't have such setup but generally should work. (should be used only for custom setups like @iruzevic has, for example it probably won't work on MAMP due to. Will implement new options for `MAMP`, `XAMPP`, `WAMP`, etc down the line)
  - Step 1: `wp config create ...`
  - Step 2: `wp core install  ...`
  - Step 3: `npm run build`
  - Step 4: `wp theme activate`

* Renaming `@since` in rename script
* Fixed the `wp db create`(now it tries doing it, but doesn't throw error if it fails). Also changed the loader text to better indicate what's happening
* "Fixed" the renaming of hiden files
* `public_html` is no longer automatically appended to `vm_dir` in VVV setup so it can be used in other Vagrant boxes (such as Scotch). Also renamed that option label from `Varying Vagrant Vagrants` to `Vagrant (VVV, Scotch, etc)`

IvanGrginovInf commented:
Changes from [feature/in-theme](https://github.com/infinum/wp-boilerplate/pull/117) have been successfully integrated with the current setup process and this **should be more or less ready to go**.

Changelog
---
* Tweaked the current `setup` script to work when everything is moved to theme
* Cut back all unneeded dependencies from project's `package.json` (in `public_html`)
* Added an export script (zips up the theme without `node-modules` or `vendors` folders - `npm run export` from theme folder)
* Added a`theme-setup script (`npm run theme-setup`) which you should run after uploading the theme to a website (but before activating it). The script does the following:
  * `npm install`
  * It prompts you for your dev-url (for browsersync) and changes the `const proxyUrl` in `webpack.config.js`
  * `composer install`
  * `npm run build`

To Do
---
* Test with other setups (non-vagrant)
* Test theme export / import on Windows

dingo-d commented on wp-content/themes/inf_theme/bin/test.js:
```diff
@@ -0,0 +1,235 @@
+  // -----------------------------
+
+  do {
+    /*
```
How come this is commented out? It's not needed or?

IvanGrginovInf replied:
@dingo-d It's a WIP (and I'm pushing because I'll have to test soon) so better if you ignore it for a while. I'll let you know when it's ready.

dingo-d commented on wp-content/themes/inf_theme/bin/test.js:
```diff
@@ -0,0 +1,235 @@
+    */
+
+    confirmed = output.summary([
+      // {label: `${emoji.get('green_book')} Theme name`, variable: newManifest.name},
```
Same here

dingo-d commented on wp-content/themes/inf_theme/bin/theme-setup.js:
```diff
@@ -76,28 +181,30 @@ const run = async() => {
-    error: 'Dev url is required and cannot be empty.',
-    required: true,
-  }).trim();
+  // const devUrl = promptFor({
```
Comments, comments everywhere 🤣

IvanGrginovInf commented:
To Do as of today:
---
* Test / fix linting - current doesn't seem to work out of the box
* Add some more "look why this is awesome" content to documentation
* Figure out naming (next meeting)
* Prepare the theme for WordPress.org (`readme.txt`, better description, screenshot, etc)
* **Issue** - Currently when you try to delete theme from WP Admin it takes a long time and then fails with "Undefined" error - need to look into this

dingo-d commented on readme.txt:
```diff
@@ -0,0 +1,106 @@
+=== Infinum WordPress Boilerplate ===
+Contributors: mustra, dingo_bastard, ivangrginov
+Tags: Starter Theme, Webpack, Node, NodeJS, npm, npx, Build Process, Autoprefixr, SCSS, SASS, LESS, Babel, Browsersync, Accessibility Ready, Custom Colors, Custom Header, Custom Logo, Custom Menu, Editor Style, Featured Images, Flexible Header, Footer Widgets, One Column, Theme Options, Translation Ready, Two Columns
```
Be very careful when adding `Accessibility Ready` tag. As any theme submitted with this tag has to go through a special review on wordpress.org.

Also, there seem to be some nonstandard tags here, which could trigger some reviewers:

[Rule about tags in style.css and description](https://make.wordpress.org/themes/handbook/review/required/#core-functionality-and-features)
[Theme tags](https://make.wordpress.org/themes/handbook/review/required/theme-tags/)

IvanGrginovInf replied:
👍 Removed most of the tags since we don't support pretty much anything in the backend without coding it yourself :)

dingo-d commented on readme.txt:
```diff
@@ -0,0 +1,106 @@
+
+= Requirements =
+
+* PHP 7 or greater (recommended: PHP 7 or greater)
```
`(recommended: PHP 7.2 or greater)` ?

IvanGrginovInf replied:
👍

dingo-d commented:
OMFG the readme! ❤️ ❤️ 😄 

I'd only change one thing: the aliasing for the phpcs. We should add to `composer.json` the `scripts` that look like this (the alias can be changed):

```json
"scripts" : {
  "check-cs": "@php ./vendor/bin/phpcs --extensions=php --ignore=/vendor/",
  "fix-cs": "@php ./vendor/bin/phpcbf --extensions=php --ignore=/vendor/",
}
```

That way you can just run (from where the composer is setup for development) 

```sh
composer check-cs .
```

Or specify the directory, and you'll have linting that's fully working. Aliasing can cause some issues, especially if you have globally installed `phpcs`- the path won't be pulled from the root of your project but the global installation and in case you are not keeping it up to date that can be an issue, also the `phpcs.xml.dist` won't be picked up from the root of your project.

Other than that really awesome work on the readme, looks sick 😄 :+1:

IvanGrginovInf commented:
Tested it and seems to work great, I'll all for it! Added the scripts and changed the readme to reflect the changes.

Regarding the `readme`, still need to add the updated Boilerplate graphics when I get them.
